sound/soc/sof/ipc4.c

```diff
@@ -0,0 +1,218 @@
+	{162, "Reserved (ADSP_IPC_PIPELINE_SAVE_FAILED removed)"},
+	{163, "Reserved (ADSP_IPC_PIPELINE_RESTORE_FAILED removed)"},
+	{165, "Reserved (ADSP_IPC_PIPELINE_ALREADY_EXISTS removed)"},
+};
+
+void sof_ipc4_check_reply_status(struct snd_sof_dev *sdev, u32 msg)
+{
+	u32 status = msg & SOF_IPC4_REPLY_STATUS_MASK;
+	int i;
+
+	sdev->ipc->msg.reply_error = status;
+	if (status) {
+		for (i = 0; i < ARRAY_SIZE(ipc4_status); i++) {
+			if (ipc4_status[i].status == status) {
+				dev_err(sdev->dev, "FW reported error: %s", ipc4_status[i].msg);
+				return;
+			}
+		}
+
+		dev_err(sdev->dev, "FW reported unknown error, status = %d", status);
+	}
+
+	snd_sof_ipc_reply(sdev, msg);
+}
+EXPORT_SYMBOL(sof_ipc4_check_reply_status);
+
+/* wait for IPC message reply */
+static int ipc4_tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
+			     void *reply_data)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	int ret;
+
+	/* wait for DSP IPC completion */
+	ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
+				 msecs_to_jiffies(sdev->ipc_timeout));
+	if (ret == 0) {
+		dev_err(sdev->dev, "error: ipc timed out for header:0x%x extension:0x%x",
+			msg->header, msg->extension);
+		return -ETIMEDOUT;
+	} else if (msg->reply_error) {
+		dev_err(sdev->dev, "error: ipc error for msg 0x%x : 0x%x\n",
+			msg->header, msg->extension);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sof_ipc4_tx_message_unlocked(struct snd_sof_ipc *ipc, u32 header,
+					u32 extension, void *msg_data, size_t msg_bytes,
+					void *reply_data, size_t reply_bytes)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	struct snd_sof_ipc_msg *msg;
+	int ret;
+
+	if (ipc->disable_ipc_tx)
+		return -ENODEV;
+
+	/*
+	 * The spin-lock is also still needed to protect message objects against
+	 * other atomic contexts.
+	 */
+	spin_lock_irq(&sdev->ipc_lock);
+
+	/* initialise the message */
+	msg = &ipc->msg;
+
+	msg->header = header;
+	msg->extension = extension;
+	msg->msg_size = msg_bytes;
+	msg->reply_size = reply_bytes;
+	msg->reply_error = 0;
+
+	/* attach any data */
+	if (msg_bytes > msg->msg_data_size)
+		return -ENOMEM;
+
+	memcpy(msg->msg_data, msg_data, msg_bytes);
+
+	sdev->msg = msg;
+
+	ret = snd_sof_dsp_send_msg(sdev, msg);
+	/* Next reply that we receive will be related to this message */
+	if (!ret)
+		msg->ipc_complete = false;
+
+	spin_unlock_irq(&sdev->ipc_lock);
+
+	if (ret < 0) {
+		dev_err_ratelimited(sdev->dev, "error: ipc tx failed with error %d", ret);
+		return ret;
+	}
+
+	/* now wait for completion */
+	if (!ret)
+		ret = ipc4_tx_wait_done(ipc, msg, reply_data);
+
+	return ret;
+}
+
+int sof_ipc4_tx_message(struct snd_sof_ipc *ipc, u32 header, u32 extension,
+			void *msg_data, size_t msg_bytes, void *reply_data,
+			size_t reply_bytes)
+{
+	int ret;
+
+	/* Serialise IPC TX */
+	mutex_lock(&ipc->tx_mutex);
+
+	ret = sof_ipc4_tx_message_unlocked(ipc, header, extension, msg_data, msg_bytes,
+					   reply_data, reply_bytes);
```

Comment:
Don't we need to change power state or at least block if we can not send IPC?
In a similar way as it was done for IPC3?

Reply:
Power setting is not supported by IPC4 now, so we can add TODO: here

Reply:
So what will happen if you try to send an IPC while the DSP is off?
Is it so that if we use IPC4 the DSP can not be powered off and it has to be always up and ready for IPC, even if there is no audio activity for hours?

Reply:
@ujfalusi no that's not the case. IPC4 does not support D0i3 atm. But D3 during runtime suspend should be fine. @bardliao correct me if I am wrong

Reply:
@ranj063 is there anything preventing D0i3 from working with IPC4? I don't understand your comment...

Reply:
> So what will happen if you try to send an IPC while the DSP is off?
> Is it so that if we use IPC4 the DSP can not be powered off and it has to be always up and ready for IPC, even if there is no audio activity for hours?

@ujfalusi yes, before power up is ready, PM is disable now, since ipc4 is still in development

Reply:
> @ranj063 is there anything preventing D0i3 from working with IPC4? I don't understand your comment...

@plbossart nothing preventing but just not tested so far. But going by Rander's comment below looks like even D3 is untested

Reply:
@ujfalusi @plbossart @ranj063 We haven't implemented PM support for IPC4 yet. We will do it in the future.

Reply:
sure. we lack pm support now. @ujfalusi do you want to join us to develop PM support for IPC4 ?

Reply:
nit-pick: it's more sof_ipc4_tx_message_locked, the lock is taken outside of the function.